src/storage/nssdb/mod.rs

```diff
@@ -1197,15 +1197,16 @@ impl Storage for NSSStorage {
                 NSS_MP_PBE_ITERATION_COUNT
             }
         };
-        let mut encdata = encrypt_data(
-            facilities,
-            newenckey.as_slice(),
-            iterations,
-            NSS_PASS_CHECK,
-        )?;
+        let mut encdata =
+            encrypt_data(facilities, &newkeys, iterations, NSS_PASS_CHECK)?;
 
         /* FIXME: need to re-encode all encrypted/integrity protected attributes */
 
+        /* now that the pin has changed all cached keys are invalid, replace the lot */
+        self.keys = newkeys;
+        /* changing the pin does not leave the token logged in */
+        self.keys.unset_key();
```

Comment:
This sounds weird. We get the new keys, we replace them in `self` and then throw away as with logout. Would not it be more straightforward just throw away both instead of doing the indirect removal of old keys by assigning new keys and then flushing them too?

Reply:
No because I want to preserve the cache of the pin which is created when we store it.

If you look carefully at the code you'll see caches are never dropped except when we whole sale replace all keys in this function, which makes sense because a change of PIN will change all keys so all caches are useless.